### PR TITLE
[ConstantFold] Fold `log1p` and `log1pf` when the input parameter is a constant value. 

### PR DESCRIPTION
**Pull request by c8ef**
This patch adds support for constant folding for the `log1p` and `log1pf` libc functions.

### CODE REVIEW

File: llvm/test/Transforms/InstCombine/log1p.ll [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+  %r = call double @log1p(double -1.000000e+00)
+  ret double %r
+}
+
+define float @log1pf_zero() {
+; CHECK-LABEL: define float @log1pf_zero() {
+; CHECK-NEXT:    [[R:%.*]] = call float @log1pf(float 0.000000e+00)
+; CHECK-NEXT:    ret float 0.000000e+00
+;
+  %r = call float @log1pf(float 0.000000e+00)
```
Comment (arsenm) [inline llvm/test/Transforms/InstCombine/log1p.ll:45]: Test -0, and other negative values less than -1 and less than 0

File: llvm/lib/Analysis/ConstantFolding.cpp [hunk 2]
```diff
@@ -2394,6 +2395,11 @@ static Constant *ConstantFoldScalarCall1(StringRef Name,
       if (!APF.isZero() && TLI->has(Func))
         return ConstantFoldFP(logb, APF, Ty);
       break;
+    case LibFunc_log1p:
+    case LibFunc_log1pf:
+      if (APF > APFloat(APF.getSemantics(), "-1") && TLI->has(Func))
```
Comment (arsenm) [inline llvm/lib/Analysis/ConstantFolding.cpp:2400]: Avoid the string? Also still fold if call is memory none

File: llvm/test/Transforms/InstCombine/log1p.ll [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+
+define double @log1p_poison() {
+; CHECK-LABEL: define double @log1p_poison() {
+; CHECK-NEXT:    [[R:%.*]] = call double @log1p(double poison)
+; CHECK-NEXT:    ret double [[R]]
+;
+  %r = call double @log1p(double poison)
+  ret double %r
+}
+
```
Comment (arsenm) [inline llvm/test/Transforms/InstCombine/log1p.ll:111]: Test memory none special cases

File: llvm/test/Transforms/InstCombine/log1p.ll [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+  %r = call float @log1pf(float 0x7FF0000000000000)
+  ret float %r
+}
+
+define double @log1p_inf() {
+; CHECK-LABEL: define double @log1p_inf() {
+; CHECK-NEXT:    [[R:%.*]] = call double @log1p(double 0x7FF0000000000000)
+; CHECK-NEXT:    ret double [[R]]
+;
+  %r = call double @log1p(double 0x7FF0000000000000)
```
Comment (arsenm) [inline llvm/test/Transforms/InstCombine/log1p.ll:72]: Test negative inf

File: llvm/test/Transforms/InstCombine/log1p.ll (hunk 3 above)
Comment (arsenm) [reply llvm/test/Transforms/InstCombine/log1p.ll:183]: I don't think the consideration of doesNotAccessMemory belongs down in ConstantFoldFP. It should produce the constant folded value regardless of the side effects. The legality of ignoring the side effects is context dependent

File: llvm/lib/Analysis/ConstantFolding.cpp (hunk 2 above)
Comment (arsenm) [reply llvm/lib/Analysis/ConstantFolding.cpp:2400]: There is an APFloat::getOne constructor

File: llvm/test/Transforms/InstCombine/log1p.ll (hunk 3 above)
Comment (arsenm) [reply llvm/test/Transforms/InstCombine/log1p.ll:183]: Plus you can and should still add the tests for these cases here, even if they do not fold yet

Comment (arsenm) [reply llvm/test/Transforms/InstCombine/log1p.ll:273]: I would consider that a defect of implementation. The separation of concerns should yield the result. If we wanted, we could produce a side piece of information for raised errors. Whether or not you care if an error occurred depend on the callsite (i.e we can ignore any exceptions in default FP environment, and memory(none) callsites can ignore errno writes)

File: llvm/lib/Analysis/ConstantFolding.cpp (hunk 2 above)
Comment (arsenm) [reply llvm/lib/Analysis/ConstantFolding.cpp:2400]: What do you mean doesn't work?

Comment (dtcxzyw) [reply llvm/lib/Analysis/ConstantFolding.cpp:2400]: `IEEEFloat::IEEEFloat(const fltSemantics &ourSemantics, integerPart value)` doesn't support initialize itself with a negative value:

https://github.com/llvm/llvm-project/blob/81fee740d073640c79c0d45a8e8e804c1c5fea40/llvm/lib/Support/APFloat.cpp#L1197-L1205
https://github.com/llvm/llvm-project/blob/a31e834ba8beaeed411b85af51ebd6598cb6007c/llvm/include/llvm/ADT/APFloat.h#L1045-L1050

Comment (arsenm) [reply llvm/lib/Analysis/ConstantFolding.cpp:2400]: So APFloat bug, cool

Comment (dtcxzyw) [reply llvm/lib/Analysis/ConstantFolding.cpp:2400]: cc @davemgreen Can you have a look and add some unittests?

Comment (dtcxzyw) [reply llvm/lib/Analysis/ConstantFolding.cpp:2400]: I will fix this.

Comment (dtcxzyw) [reply llvm/lib/Analysis/ConstantFolding.cpp:2400]: Fixed. Please rebase this patch on https://github.com/llvm/llvm-project/commit/a54d88f97dfeb35e4f6c14e5563ec686e68fd244.

**Review by arsenm — APPROVED** [review]

**Comment by jakeegan** [comment]
This test fails on the AIX bot, could you take a look?
https://lab.llvm.org/buildbot/#/builders/64/builds/1232/steps/6/logs/FAIL__LLVM__log1p_ll

**Comment by jakeegan** [comment]
@c8ef The result is:
```
define noundef double @f() local_unnamed_addr #0 {
entry:
  ret double 0.000000e+00
}
```
It seems this will require more investigation on an AIX machine, so I'll XFAIL the test on AIX and investigate further.

**Comment by hubert-reinterpretcast** [comment]
The negative zero behaviour is optional. Clang on AIX defines neither `__STDC_IEC_559__` nor `__STDC_IEC_60559_BFP__`.

**Comment by jakeegan** [comment]
@c8ef Would you be able to adapt the test to reflect negative zero behaviour being optional?

**Comment by hubert-reinterpretcast** [comment]
> Marking it as `XFAIL: target={{.*}}-aix{{.*}}` will be sufficient?

The XFAIL is actually wrong. The real issue is that compiling for a non-AIX target on an AIX machine will not generate the negative zero result.

A possible solution is to add the special cases to the constant-folding code in LLVM.

**Comment by arsenm** [comment]
Can you write a regex check allowing flexibility in the low bits?

**Comment by arsenm** [comment]
Also could special case the zero

**Comment by hubert-reinterpretcast** [comment]
> Also could special case the zero

https://github.com/llvm/llvm-project/pull/114635 created to implement that direction.

**Pull request merged**